File: 10-Adapter/src/go/stripe_api.go
package main

import (
	"fmt"
	"time"
)

// Third-party Stripe API — different interface, amounts in cents.
type StripeApi struct{}

type StripeResult struct {
	Success bool
	ID      string
}

func (s *StripeApi) CreateCharge(cents int, currency, token string) StripeResult {
	fmt.Printf("  [Stripe] Charging %d cents (%s) token=%s\n", cents, currency, token)
	return StripeResult{Success: true, ID: fmt.Sprintf("ch_%d", time.Now().UnixNano())}
}

func (s *StripeApi) CreateRefund(chargeId string) StripeResult {
	fmt.Printf("  [Stripe] Refunding charge %s\n", chargeId)
	return StripeResult{Success: true, ID: fmt.Sprintf("re_%d", time.Now().UnixNano())}
}

File: 10-Adapter/src/rust/stripe_api.rs
/// Third-party Stripe API — different interface, amounts in cents.
pub struct StripeApi;

pub struct StripeResult {
    pub success: bool,
    pub id: String,
}

impl StripeApi {
    pub fn create_charge(&self, cents: i32, currency: &str, token: &str) -> StripeResult {
        println!("  [Stripe] Charging {} cents ({}) token={}", cents, currency, token);
        StripeResult {
            success: true,
            id: format!("ch_{}", std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH).unwrap().as_nanos()),
        }
    }

    pub fn create_refund(&self, charge_id: &str) -> StripeResult {
        println!("  [Stripe] Refunding charge {}", charge_id);
        StripeResult {
            success: true,
            id: format!("re_{}", std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH).unwrap().as_nanos()),
        }
    }
}
